### ISSUE TITLE
Nelua and C disagree on type size on emscripten

### ISSUE THREAD

**Issue by Andre-LA** [edubart/nelua-game2048#1]
Using:
`nelua game2048.nelua --cc emcc`

I get these C compile errors:

```
nelua_cache/game2048.c:100:1: error: static_assert failed due to requirement 'sizeof(struct nelua_GCItem) == 40' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nelua_GCItem) == 40, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:110:1: error: static_assert failed due to requirement 'sizeof(struct nelua_span_GCItem) == 16' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nelua_span_GCItem) == 16, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:122:1: error: static_assert failed due to requirement 'sizeof(struct nlstringview) == 16' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nlstringview) == 16, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:142:1: error: static_assert failed due to requirement 'sizeof(struct nelua_span_uint8) == 16' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nelua_span_uint8) == 16, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:170:1: error: static_assert failed due to requirement 'sizeof(struct game2048_span_TileAnim) == 16' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(game2048_span_TileAnim) == 16, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:220:1: error: static_assert failed due to requirement 'sizeof(struct nelua_GC) == 104' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nelua_GC) == 104, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:234:1: error: static_assert failed due to requirement 'sizeof(nlusize_arr24) == 192' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nlusize_arr24) == 192, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:332:1: error: static_assert failed due to requirement 'sizeof(struct nelua_stringbuffer) == 24' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nelua_stringbuffer) == 24, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:345:1: error: static_assert failed due to requirement 'sizeof(struct nlstring) == 16' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(nlstring) == 16, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:399:1: error: static_assert failed due to requirement 'sizeof(struct game2048_vector_TileAnim) == 24' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(game2048_vector_TileAnim) == 24, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:438:1: error: static_assert failed due to requirement 'sizeof(game2048_vector_TileAnim_arr4) == 96' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(game2048_vector_TileAnim_arr4) == 96, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:440:1: error: static_assert failed due to requirement 'sizeof(game2048_AnimTileMap) == 384' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(game2048_AnimTileMap) == 384, "Nelua and C disagree on type size");
^                   ~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
nelua_cache/game2048.c:23:29: note: expanded from macro 'nelua_static_assert'
#define nelua_static_assert _Static_assert
                            ^
nelua_cache/game2048.c:451:1: error: static_assert failed due to requirement 'sizeof(struct record_3verfCjmM8wbHgWWfabDjo) == 24' "Nelua and C disagree on type size"
nelua_static_assert(sizeof(record_3verfCjmM8wbHgWWfabDjo) == 24, "Nelua and C disagree on type size");
```

**Comment by edubart** [comment]
Use --cpu-bits=32

**Comment by Andre-LA** [comment]
I got this error:

```
/usr/local/lib/luarocks/rocks-5.3/nelua/dev-1/conf/lib/stringview.nelua:91:41: error: in binary operation `add`: invalid operation between types 'usize' and 'isize'
  if unlikely(init < 0) then init = self.size + init + 1 end
```

**Comment by edubart** [comment]
I can't reproduce this here, how you got that error?

**Comment by Andre-LA** [comment]
I don't know exactly.

with `nelua game2048.nelua`, the game runs fine.
but with `nelua game2048.nelua --cpu-bits=32`, I got this error on stringview.nelua

I'm using Lua 5.3

**Comment by edubart** [comment]
Are you using latest Nelua compiler?

**Comment by Andre-LA** [comment]
yes

**Comment by edubart** [comment]
I've just compiled this repo again using emscripten, with the flags:
Used this command
```
nelua --cc=emcc --cpu-bits=32 --compile-binary game2048.nelua
```
I made some fixes in the Nelua compiler to handle 32 bit CPUs better, might help.
I've also updated the `config.nelua` file in repository to include the link raylib correctly in emscripten.

**Comment by Andre-LA** [comment]
updated Nelua and game repository, got this new error:

```
$ nelua --cc=emcc --cpu-bits=32 --compile-binary game2048.nelua
/usr/local/lib/luarocks/rocks-5.3/nelua/dev-1/conf/lib/stringview.nelua:94:62: error: in binary operation `sub`: invalid operation between types 'usize' and 'isize'
  local found: pointer = memory.find(&self.data[init-1], self.size-init+1, pattern.data, pattern.size)
                                                             ^
```

**Comment by edubart** [comment]
I am getting this error now only when using Lua 5.3, not with Lua 5.4, will investigate and fix..

**Comment by edubart** [comment]
I've fixed in Nelua, the problem was because Lua 5.4 was printing a number as `64` while Lua 5.3 was printing as `64.0`.

**Comment by Andre-LA** [comment]
It works!
I still got an error (raylib.h file not found), but it's about my raylib+emscripten installation.
